plaster-yaml broken for Python < v10 due to importlib.metadata API changes

Hi,

As mentioned in the thread of  PR #3 (quoted at bottom) the importlib.metadata API has not been stable and the current code uses importlib.metadata.entry_points() features that are not available until Python 3.10.

I tested this by running the regression tests using 3.9.  (See below.)
You can see here <https://github.com/python/importlib_metadata> that Python 3.8 is using the same code as 3.9 so 3.8 is also broken.

If you don't like the 11 lines of code in PR #3 that does the import, you could look at plaster.  It gets away with 4 lines to do the import. (See: <https://github.com/Pylons/plaster/blob/1cd0658d2b7c86325815f0cff3d4c8a6c9df11ff/src/plaster/loaders.py#L1>).  But then it pays for that brevity with another 17 liines of code to work-around issues solved in later versions of importlib.metadata. (See: <https://github.com/Pylons/plaster/blob/1cd0658d2b7c86325815f0cff3d4c8a6c9df11ff/src/plaster/loaders.py#L169>)  (At least it seems at first glance that the problems are solved in later versions of importlib.metadata.  I didn't spend the time to be sure.)

To reproduce I installed pyenv.
```
cd ${HOME}
mkdir -p .pyenv
cd .pyenv
git clone https://github.com/pyenv/pyenv.git
```

I then installed Python 3.9.18, found at <https://www.python.org/downloads/>.  (You could try 3.9.19.  I don't know why I didn't use that.)

After un-tar-ing and cd-ing into the source directory:
```
mkdir -p ~/.pyenv/versions/3.9.18
./configure --prefix=${HOME}/.pyenv/versions/3.9.18/
make
make test
make install
```

I used pyenv to run Python 3.9.18 as follows, but you can do as you like:
```
bash  ; # Start a new shell, so I can exit and go back to normal
PATH=${HOME}/.pyenv/bin/:${PATH}
# What follows is what `pyenv init` says to do/put in your ~/.bashrc
export PYENV_ROOT="$HOME/.pyenv"
[[ -d $PYENV_ROOT/bin ]] && export PATH="$PYENV_ROOT/bin:$PATH"
eval "$(pyenv init -)"
# Test that the following says version 3.9.18
python3 --version
```
Then go run the plaster-yaml regression tests.

I got the following output (4 tests failed):
```
$ ../venv9/bin/pytest
============================= test session starts ==============================
platform linux -- Python 3.9.18, pytest-6.2.5, py-1.11.0, pluggy-1.4.0
rootdir: /home/kop/projects/pyramid/pyramid-cookbook/plaster-yaml/plaster-yaml
collected 9 items                                                              

tests/test_plaster_yaml.py FF.....FF                                     [100%]

=================================== FAILURES ===================================
____________________________ test_resolve_waitress _____________________________

    def test_resolve_waitress():
>       main = resolve_use("egg:waitress#main", "paste.server_runner")

tests/test_plaster_yaml.py:14: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

use = 'egg:waitress#main', entrypoint = 'paste.server_runner'

    def resolve_use(use: str, entrypoint: str) -> Callable:
        try:
            pkg, name = use.split("#")
        except ValueError:
            pkg, name = use, "main"
        try:
            scheme, pkg = pkg.split(":")
        except ValueError:
            scheme = "egg"
        if scheme != "egg":
            raise ValueError(f"{use}: unsupported scheme {scheme}")
    
>       eps = metadata.entry_points(group=entrypoint, name=name)
E       TypeError: entry_points() got an unexpected keyword argument 'group'

src/plaster_yaml/loader.py:43: TypeError
____________________________ test_resolve_gunicorn _____________________________

    def test_resolve_gunicorn():
>       main = resolve_use("egg:gunicorn#main", "paste.server_runner")

tests/test_plaster_yaml.py:19: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

use = 'egg:gunicorn#main', entrypoint = 'paste.server_runner'

    def resolve_use(use: str, entrypoint: str) -> Callable:
        try:
            pkg, name = use.split("#")
        except ValueError:
            pkg, name = use, "main"
        try:
            scheme, pkg = pkg.split(":")
        except ValueError:
            scheme = "egg"
        if scheme != "egg":
            raise ValueError(f"{use}: unsupported scheme {scheme}")
    
>       eps = metadata.entry_points(group=entrypoint, name=name)
E       TypeError: entry_points() got an unexpected keyword argument 'group'

src/plaster_yaml/loader.py:43: TypeError
_____________________________ test_get_wsgi_server _____________________________

serve_paste = <MagicMock name='serve_paste' id='139692513139872'>
loader = <plaster_yaml.loader.Loader object at 0x7f0cb2991ee0>

    @patch("waitress.serve_paste")
    @pytest.mark.usefixtures("loader")
    def test_get_wsgi_server(serve_paste, loader):
>       wsgi_server = loader.get_wsgi_server()

tests/test_plaster_yaml.py:75: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
src/plaster_yaml/loader.py:93: in get_wsgi_server
    server = resolve_use(settings.pop("use"), "paste.server_runner")
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

use = 'egg:waitress#main', entrypoint = 'paste.server_runner'

    def resolve_use(use: str, entrypoint: str) -> Callable:
        try:
            pkg, name = use.split("#")
        except ValueError:
            pkg, name = use, "main"
        try:
            scheme, pkg = pkg.split(":")
        except ValueError:
            scheme = "egg"
        if scheme != "egg":
            raise ValueError(f"{use}: unsupported scheme {scheme}")
    
>       eps = metadata.entry_points(group=entrypoint, name=name)
E       TypeError: entry_points() got an unexpected keyword argument 'group'

src/plaster_yaml/loader.py:43: TypeError
______________________________ test_get_wsgi_app _______________________________

loader = <plaster_yaml.loader.Loader object at 0x7f0cb2991520>

    @pytest.mark.usefixtures("loader")
    def test_get_wsgi_app(loader):
>       wsgi_app = loader.get_wsgi_app()

tests/test_plaster_yaml.py:91: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
src/plaster_yaml/loader.py:98: in get_wsgi_app
    use = resolve_use(settings.pop("use"), "paste.app_factory")
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

use = 'egg:pyramid_helloworld', entrypoint = 'paste.app_factory'

    def resolve_use(use: str, entrypoint: str) -> Callable:
        try:
            pkg, name = use.split("#")
        except ValueError:
            pkg, name = use, "main"
        try:
            scheme, pkg = pkg.split(":")
        except ValueError:
            scheme = "egg"
        if scheme != "egg":
            raise ValueError(f"{use}: unsupported scheme {scheme}")
    
>       eps = metadata.entry_points(group=entrypoint, name=name)
E       TypeError: entry_points() got an unexpected keyword argument 'group'

src/plaster_yaml/loader.py:43: TypeError
=============================== warnings summary ===============================
../venv9/lib/python3.9/site-packages/paste/deploy/loadwsgi.py:8
  /home/kop/projects/pyramid/pyramid-cookbook/plaster-yaml/venv9/lib/python3.9/site-packages/paste/deploy/loadwsgi.py:8: DeprecationWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html
    import pkg_resources

../venv9/lib/python3.9/site-packages/pkg_resources/__init__.py:2832
  /home/kop/projects/pyramid/pyramid-cookbook/plaster-yaml/venv9/lib/python3.9/site-packages/pkg_resources/__init__.py:2832: DeprecationWarning: Deprecated call to `pkg_resources.declare_namespace('paste')`.
  Implementing implicit namespace packages (as specified in PEP 420) is preferred to `pkg_resources.declare_namespace`. See https://setuptools.pypa.io/en/latest/references/keywords.html#keyword-namespace-packages
    declare_namespace(pkg)

../venv9/lib/python3.9/site-packages/pkg_resources/__init__.py:2832
../venv9/lib/python3.9/site-packages/pkg_resources/__init__.py:2832
  /home/kop/projects/pyramid/pyramid-cookbook/plaster-yaml/venv9/lib/python3.9/site-packages/pkg_resources/__init__.py:2832: DeprecationWarning: Deprecated call to `pkg_resources.declare_namespace('zope')`.
  Implementing implicit namespace packages (as specified in PEP 420) is preferred to `pkg_resources.declare_namespace`. See https://setuptools.pypa.io/en/latest/references/keywords.html#keyword-namespace-packages
    declare_namespace(pkg)

-- Docs: https://docs.pytest.org/en/stable/warnings.html
=========================== short test summary info ============================
FAILED tests/test_plaster_yaml.py::test_resolve_waitress - TypeError: entry_p...
FAILED tests/test_plaster_yaml.py::test_resolve_gunicorn - TypeError: entry_p...
FAILED tests/test_plaster_yaml.py::test_get_wsgi_server - TypeError: entry_po...
FAILED tests/test_plaster_yaml.py::test_get_wsgi_app - TypeError: entry_point...
=================== 4 failed, 5 passed, 4 warnings in 0.57s ====================
```


----
             On Sat, 30 Mar 2024 01:40:37 -0700
Guillaume Gauvrit ***@***.***> wrote:

> @mardiros commented on this pull request.
> 
> 
> 
> >  description = "A plaster plugin to configure pyramid app with
> > Yaml"  
>  readme = "README.md"
>  authors = ["Guillaume Gauvrit ***@***.***>"]
>  include = ["CHANGELOG.md"]
>  license = "BSD-derived"
>  
>  [tool.poetry.dependencies]
> +# See: https://pypi.org/project/backports.entry-points-selectable
> +importlib_metadata = { version = ">=3.6", python = "<3.10" }
> 
> I was waiting to drop the python 3.7 support to remove pkg_resources
> usage.
> 
> Is it ok for you to drop python 3.7 support ?
> 
> importlib.metadata is avaible in python 3.8
> so updating the ligne 13 with  the code below should be enough.
> 
> ```
> python = "^3.8"
> ```

Unfortunately it is not enough.  The entry_points() API did
not stabilize until python 3.10.  See:
https://pypi.org/project/backports.entry-points-selectable/
So you have to use a backport if you want to use the same code
in 3.8 and 3.9 as in 3.10.  In that case you may as well
support 3.7.

Also, take a look at the last paragraphs of:
https://docs.python.org/3/library/importlib.metadata.html#entry-points

I'm running python 3.11, and used it to successfully run the regression
tests.  So the existing code works, but it's not certain that
it will continue to work on 3.12.  My reading of the importlib.metadata
entry point docs above lead me to believe that EntryPoints is
an iterable object.  But I'm not all that steeped in Python's
vocabulary so I'm not 100% certain I'm right.

(I suppose I could try reading the source.  It's probably
easier to ask in #python.  :-)

> > @@ -1,5 +1,15 @@  
> -import pkg_resources
> +import sys
> +
> +if sys.version_info < (3, 10):
> +    import importlib_metadata
> +
> +    class importlib:
> +        pass
> +
> +    setattr(importlib, "metadata", importlib_metadata)
> +else:
> +    import importlib.metadata
> 
> this is why I dislike that. 11 lines  of code to retrieve a version
> number. this is outstanding.

Agreed.  But at least it's an easy-to-maintain 11 lines.  It's clear
what needs to be done and when it's time to do it.

> please remove python 3.7 support and everything will be much simpler.

Let me know how you'd like to proceed. 

Regards,

Karl ***@***.***>
Free Software:  "You don't pay back, you pay forward."
                 -- Robert A. Heinlein

_Originally posted by @kpinc in https://github.com/mardiros/plaster-yaml/issues/3#issuecomment-2028011665_

--- a/.github/workflows/main.yml
+++ b/.github/workflows/main.yml
@@ -21,7 +21,7 @@ jobs:
     runs-on: ubuntu-latest
     strategy:
       matrix:
-        python-version: ["3.7", "3.8", "3.9", "3.10", "3.11"]
+        python-version: ["3.8", "3.9", "3.10", "3.11"]
         poetry-version: ["1.3.1"]
     # Steps represent a sequence of tasks that will be executed as part of the job
     steps:
@@ -58,10 +58,10 @@ jobs:
           name: pytest-results-${{ matrix.python-version }}
           path: junit/test-results-${{ matrix.python-version }}.xml
 
-      # - name: Codecov
-      #   uses: codecov/codecov-action@v3
-      #   env:
-      #     CODECOV_TOKEN: ${{ secrets.CODECOV_TOKEN }}
-      #   with:
-      #     # Comma-separated list of files to upload
-      #     files: coverage.xml
+      - name: Codecov
+        uses: codecov/codecov-action@v3
+        env:
+          CODECOV_TOKEN: ${{ secrets.CODECOV_TOKEN }}
+        with:
+          # Comma-separated list of files to upload
+          files: coverage.xml

--- a/.github/workflows/main.yml
+++ b/.github/workflows/main.yml
@@ -59,7 +59,7 @@ jobs:
           path: junit/test-results-${{ matrix.python-version }}.xml
 
       - name: Codecov
-        uses: codecov/codecov-action@v3
+        uses: codecov/codecov-action@v4
         env:
           CODECOV_TOKEN: ${{ secrets.CODECOV_TOKEN }}
         with:

--- a/README.md
+++ b/README.md
@@ -2,6 +2,7 @@
 
 Configure your applications with YAML (and JSON) structured data.
 
+[![CI](https://github.com/mardiros/plaster-yaml/actions/workflows/main.yml/badge.svg)](https://github.com/mardiros/plaster-yaml/actions/workflows/main.yml)
 [![codecov](https://codecov.io/gh/mardiros/plaster-yaml/graph/badge.svg?token=LKGPH2OJDN)](https://codecov.io/gh/mardiros/plaster-yaml)
 
 ## Introduction

--- a/README.md
+++ b/README.md
@@ -2,6 +2,8 @@
 
 Configure your applications with YAML (and JSON) structured data.
 
+[![codecov](https://codecov.io/gh/mardiros/plaster-yaml/graph/badge.svg?token=LKGPH2OJDN)](https://codecov.io/gh/mardiros/plaster-yaml)
+
 ## Introduction
 
 The `plaster-yaml` plugin lets applications, notably [Pyramid](

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -11,8 +11,8 @@ license = "BSD-derived"
 
 [tool.poetry.dependencies]
 python = "^3.8"
+importlib-metadata = { version = "^7.1.0", python = "<3.10" }
 plaster = "^1.0"
-# Upgraded PyYAML from ^5.4 due to: https://github.com/yaml/pyyaml/issues/601
 PyYAML = "^6.0.1"
 
 [tool.poetry.dev-dependencies]

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -25,6 +25,9 @@ pytest = "^6.2.5"
 waitress = "^2.0.0"
 pyramid-helloworld = "^1.1.0"
 
+[tool.poetry.group.dev.dependencies]
+pytest-cov = "^5.0.0"
+
 [build-system]
 requires = ["poetry-core>=1.0.0"]
 build-backend = "poetry.core.masonry.api"

--- a/src/plaster_yaml/__init__.py
+++ b/src/plaster_yaml/__init__.py
@@ -1,6 +1,5 @@
-from importlib import metadata
-
 from .loader import Loader
+from .compat import importlib_metadata
 
 __all__ = ["Loader"]
-__version__ = metadata.version("plaster-yaml")
+__version__ = importlib_metadata.version("plaster-yaml")

--- a/src/plaster_yaml/compat.py
+++ b/src/plaster_yaml/compat.py
@@ -1,0 +1,8 @@
+import sys
+
+if sys.version_info >= (3, 10):
+    from importlib import metadata as importlib_metadata
+else:
+    import importlib_metadata
+
+__all__ = ["importlib_metadata"]

--- a/src/plaster_yaml/loader.py
+++ b/src/plaster_yaml/loader.py
@@ -3,7 +3,7 @@ import pathlib
 from logging.config import dictConfig
 from typing import Callable
 
-from importlib import metadata
+from .compat import importlib_metadata
 
 import plaster
 import yaml
@@ -40,7 +40,7 @@ def resolve_use(use: str, entrypoint: str) -> Callable:
     if scheme != "egg":
         raise ValueError(f"{use}: unsupported scheme {scheme}")
 
-    eps = metadata.entry_points(group=entrypoint, name=name)
+    eps = importlib_metadata.entry_points(group=entrypoint, name=name)
     (runner,) = [ep for ep in eps if ep.module.split(".")[0] == pkg]
     return runner.load()
 

